feat: agent toolkit library

dtbuchholz opened the pull request:
## Summary

This adds a new `@recallnet/agent-toolkit`, which brings framework agnostic agentic flows to Recall. Closes https://github.com/recallnet/js-recall/issues/115

## Details

Heavily inspired by the [Stripe agent toolkit](https://github.com/stripe/agent-toolkit), it provides:
- A set of `shared` tools, prompts, and API to the Recall network
- Framework adapters for OpenAI, AI SDK, LangChain, and MCP—all which use the same shared tools but have slightly different dependencies and instantiation logic etc
- Builds (but doesnt export) a set of examples for each framework so that devs can run them
   - Note: I'm not sure if this is the right pattern, or if we should just have separate repos/docs and remove the `examples`. Counterpoint: having them in the package helped with testing the adapters work properly. 

The best way to see how these work is to review the README, or install/build and run `node dist/examples/...` for one of the packages. The MCP example is probably the easiest to inspect via: `SERVER_PORT=3001 npx @modelcontextprotocol/inspector node dist/examples/mcp.js --tools=all --network=localnet`

## Notes

if you run `pnpm build`, the `dist/examples` have some prebuilt programs you can run, including MCP. we might want to move the related deps to dev deps.

from a packaging perspective, the one callout is that i ended up with a barrel file export for the `shared` module (`@recallnet/agent-toolkit/shared`). i can split out the paths to be more explicit, but the way i see this package being used is in two ways:
1. youre working with a framework/tools like LangChain, OAI, AI SDK, or MCP—and you want to easily import prebuilt Recall tools that agents can use w/o much effort. e.g., `@recallnet/agent-toolkit/mcp`
2. you want to have greater control and implement things on your own—which is you want to use the `shared` deps.  e.g., `@recallnet/agent-toolkit/shared`

since i think most ppl will opt for the flow in (1), i decided the barrel file export might be easier so that there aren't too many different paths to choose from? idk or maybe it's fine to replace `@recallnet/agent-toolkit/shared` with:
- `@recallnet/agent-toolkit/api`
- `@recallnet/agent-toolkit/configuration`
- `@recallnet/agent-toolkit/functions`
- `@recallnet/agent-toolkit/parameters`
- `@recallnet/agent-toolkit/prompts`
- `@recallnet/agent-toolkit/tools`
- `@recallnet/agent-toolkit/util`

dtbuchholz commented:
i have to make some tweaks with the new requirements in https://github.com/recallnet/js-recall/issues/115. mainly:
1. split out the `examples/mcp.ts` into its own package at `packages/mcp`, and combine with https://github.com/recallnet/recall-mcp/. this probably makes sense to do in a separate PR, though.
2. general refactoring—but i think it'd be nice to make any additional tweaks based on feedback, like do the APIs make or the resource/permission patterns make sense? (since it copied Stripe, i actually quite like the way it's handled.)

dtbuchholz commented:
@carsonfarmer just wanted to flag this PR for a proper review. i can do a bit more refactoring before i officially open it, but before doing so, i wanted to get a gut check on the current design. 

i added some notes in the PR description as well as in the comment above.

dtbuchholz commented:
note: for whatever feedback you have, plus the general refactoring & tests—i'll open a secondary PR that points to this one, making it easier to review any final changes. this can also include the `packages/mcp` setup that i mentioned above.

carsonfarmer commented on packages/agent-toolkit/README.md:
````diff
@@ -0,0 +1,231 @@
+Install the package:
+
+```bash
+npm install @recallnet/agent-toolkit
````
Do we want to also have pnpm install steps? They aren't really all that different. Additionally, does this work with deno? Probably would via `deno add npm:@recallnet/agent-toolkit`..?

carsonfarmer commented on packages/agent-toolkit/README.md:
```diff
@@ -0,0 +1,231 @@
+
+The permissions are:
+
+- `read`: Read-only operations (i.e., do not cost any tokens/credit)
```
```suggestion
- `read`: Read-only operations (i.e., operations do not cost any tokens/credit)
```

carsonfarmer commented on packages/agent-toolkit/README.md:
```diff
@@ -0,0 +1,231 @@
+The permissions are:
+
+- `read`: Read-only operations (i.e., do not cost any tokens/credit)
+- `write`: Write operations, which modify state (i.e., submit transactions)
```
```suggestion
- `write`: Write operations, which modify state (i.e., operations may submit transactions)
```

carsonfarmer commented on packages/agent-toolkit/README.md:
````diff
@@ -0,0 +1,231 @@
+```ts
+import { Configuration } from "@recallnet/agent-toolkit/shared";
+
+const config: Configuration = {
````
Is this the full set of config options at the moment?

dtbuchholz replied:
yeah, i kept it simple in case we want to enhance this in the future. if you expand it, it looks like this. the config lets you define:
- the "action" groups and permissions you want to enable (like `bucket` related scopes with `read` only functionality)
- any additional context that can be used by the actual tool calls. 
```ts
export type Resource = "account" | "bucket";
export type Permission = "read" | "write";
export type Configuration = {
  actions: {
    [K in Resource]?: {
      [K in Permission]?: boolean;
    };
  };
  context: {
    network?: string;
    [key: string]: any;
  };
};
```
we dont _really_ use any special context in actual tool calls, but i wanted to give us the option to do so. the existing `context` has a `network` field, which only is used upon the `RecallAPI` instantiation.
```ts
export default class RecallAPI {
  constructor(
    privateKey: string,
    context?: Context,
    serializer: (data: any) => string = jsonStringify,
  ) {
    // Use the `context.network` string to set up the connected chain
    const chain =
      context?.network !== undefined &&
      checkChainName(context.network as ChainName)
        ? getChain(context.network as ChainName)
        : testnet;
    const walletClient = walletClientFromPrivateKey(privateKey as Hex, chain);
    const recallClient = new RecallClient({ walletClient });

    this._recall = recallClient;
    this._context = context || {};
    this._serialize = serializer;
  }
```
so, how can we better use the `Configuration`? the `context` is passed to every tool function. e.g., a few methods use a `params.bucket` parameter and pass that address to the JS SDK method call. instead, we could choose to alter that flow such that the `context` provides the param as a "global" variable. here's what the current state looks like—you can see that `context` isnt used at all:
```ts
export const addObject = async (
  recall: RecallClient,
  context: Context,
  params: z.infer<typeof addObjectParameters>,
): Promise<Result<{ txHash: string }>> => {
  try {
    const metadata = params.metadata ?? {};
    const dataToStore =
      typeof params.data === "string"
        ? new TextEncoder().encode(params.data)
        : params.data;

    const { meta } = await recall
      .bucketManager()
      .add(params.bucket as Address, params.key, dataToStore, {
        overwrite: params.overwrite ?? false,
        metadata: { ...metadata },
      });
    if (!meta?.tx) {
      return { success: false, error: "Transaction not found" };
    }
    return { success: true, result: { txHash: meta.tx.transactionHash } };
  } catch (error: any) {
    return { success: false, error: `Failed to add object: ${error.message}` };
  }
};
```
e.g., we could enhance the config to let you:
1. you set a global bucket
2. when you call `addObject`, we make the `params.bucket` optional; if you set a global bucket, we'll use that unless overridden by `params.bucket`
```ts
export const addObject = async (
  recall: RecallClient,
  context: Context,
  params: z.infer<typeof addObjectParameters>,
): Promise<Result<{ txHash: string }>> => {
    let bucket;
    if (context.bucket) bucket = context.bucket;
    if (params.bucket) bucket = params.bucket;
    if (bucket === undefined) throw new Error("No bucket set in context or passed to method");
    // ...
}

carsonfarmer commented on packages/agent-toolkit/package.json:
```diff
@@ -0,0 +1,87 @@
+    "@recallnet/sdk": "workspace:*",
+    "ai": "^4.1.61",
+    "dotenv": "^16.4.7",
+    "langchain": "^0.3.19",
```
Is there any way we can move some of these to peer dependencies? Or is code from these packages actually used directly within our package itself?

https://nodejs.org/en/blog/npm/peer-dependencies

dtbuchholz replied:
yeh, i'll move them—a handful arent actually dependences and are here because they're used in the "examples" module. instead, we should split them out into dev deps, i think, since they're not actually exported and are here for demo purposes

carsonfarmer commented on packages/agent-toolkit/src/ai-sdk/tool.ts:
```diff
@@ -0,0 +1,27 @@
+  description: string,
+  schema: z.ZodObject<any, any, any, any, { [x: string]: any }>,
+): CoreTool {
+  return tool({
```
ah ok, so this probably answers my previous question

carsonfarmer replied:
oh no wait, this would still work right? to mark ai-ask as a peer dep?

carsonfarmer commented on packages/agent-toolkit/src/ai-sdk/toolkit.ts:
```diff
@@ -0,0 +1,89 @@
+    });
+  }
+
+  middleware(): LanguageModelV1Middleware {
```
can you explain why we need this? just because of the framework agnostic factor? because in this case, we're not _actually_ providing any "middleware" functionality, we're just passing things right along...

carsonfarmer commented on packages/agent-toolkit/src/examples/ai-sdk.ts:
```diff
@@ -0,0 +1,72 @@
+  },
+});
+
+const model = wrapLanguageModel({
```
so, this doesn't _actually_ do anything, because our toolkit doesn't actually provide any middleware functionality, right?

dtbuchholz replied:
true. i sorta just copied the Stripe toolkit, which had a bunch of custom middleware logic that tracked Stipe account billing. let me see if i can just remove this middleware altogether.

carsonfarmer commented on packages/agent-toolkit/src/examples/langchain.ts:
```diff
@@ -0,0 +1,86 @@
+import { dirname, resolve } from "path";
+import { fileURLToPath } from "url";
+
+import { RecallAgentToolkit } from "../langchain/index.js";
```
can we set things up so that the import is actually the public api import? I think we can do that with aliases right? so it'd be `import { RecallAgentToolkit } from "@recallnet/agent-toolkit/langchain"` here?

carsonfarmer commented on packages/agent-toolkit/src/examples/mcp.ts:
```diff
@@ -0,0 +1,182 @@
+#!/usr/bin/env node
```
This is great, though as you point it, it'd be nice if we just made this its own sub-package so folks can just use it not as an example, but rather as a standalone mcp implementation, which _also_ acts as an example!

Note also that this (https://github.com/wong2/litemcp/) is a pretty nice framework for building mcp services very easily.

carsonfarmer commented on packages/agent-toolkit/src/shared/parameters.ts:
```diff
@@ -0,0 +1,172 @@
+export const getAccountInfoParameters = z.object({
+  address: z
+    .string()
+    .regex(/^0x[a-fA-F0-9]{40}$/)
```
should probably make this into a const

carsonfarmer commented on packages/agent-toolkit/src/shared/parameters.ts:
```diff
@@ -0,0 +1,172 @@
+import { z } from "zod";
```
lovely module!

carsonfarmer replied:
I didn't double check your zod defs here against any official source of truth... do we _have_ an official source of truth for these from which we can derive the zod schemas? Like json schema docs?

carsonfarmer commented on packages/agent-toolkit/src/shared/prompts.ts:
```diff
@@ -0,0 +1,72 @@
+export const getAccountInfoPrompt = `
```
nice, again, wondering if we should have a single source of truth for these (maybe this is it), but that is language agnostic? like just adding it to the docs string for a json schema definition?

carsonfarmer replied:
what is best practices here?

dtbuchholz replied:
yeah i guess we'd need a language-agnostic spec. perhaps the scope for the agent memory specs https://github.com/recallnet/competition-v0/issues/1 could be expanded to incorporate this idea?

or, since the MCP spec uses JSON RPC 2.0, we could create a JSON RPC spec for all of these prompts etc—and then import that spec into the agent toolkit https://spec.modelcontextprotocol.io/specification/2024-11-05/server/prompts/#getting-a-prompt.

carsonfarmer replied:
Can you create a ticket for this in that repo pls?

carsonfarmer commented on packages/agent-toolkit/tsup.config.ts:
```diff
@@ -0,0 +1,87 @@
+      js: format === "esm" ? ".js" : ".cjs",
+    }),
+  },
+  // Also, publish examples so that they can be used as standalone processes (e.g., MCP server)
```
For the mcp server, this makes sense. Maybe not for the others though?

carsonfarmer reviewed (approved):
This is looking really solid to me @dtbuchholz. I'm giving this a 👍 to finalize and move forward. Then we'll get a proper review from myself and maybe @mzkrasner.

dtbuchholz commented:
okay, i think i addressed your feedback @carsonfarmer:
- clean up the package exports and dependencies:
   - move things into an `examples` dir outside of `src`
   - set up optional peer or dev deps accordingly
- linting fixes for `any` types (mostly, ignoring them)
- clean up the ai-sdk adapter/example (removed the middleware) and MCP example (default to all `--tools` if not provided)
- make the zod address/regex type a constant so that it's reused across the different method parameter schemas
- docs cleanups for the install steps and examples

carsonfarmer reviewed (approved):
Yup, agreed. It looks like you've addressed my comments. Let's get this merged and ready to use asap. Can we publish pre-release versions of these libs so that we can use the in other "internal" projects for now?

carsonfarmer commented:
Also, let's do a fast follow (as time permits) on the standalone mcp service with proper "bin" tooling, so we can do `npx @reacallnet/agent-toolkit/mcp` in mcp configs etc.

dtbuchholz commented:
> Can we publish pre-release versions of these libs so that we can use the in other "internal" projects for now?

We use `changesets` for versioning/publishing, which does have a prerelease versioning feature. I'd have to see how that works with the monorepo, though, since it might conflict with the other packages that don't need it. Worst case, I'll just publish it as `0.0.1`. 

I'll merge this PR as-is. Then, I'll set up the separate MCP package/binary, and within that PR, I'll include what we need to version/release both the agent toolkit and the MCP server. I should be able to get that done before tmrw.
